// @mapbox/search-js-react/web/Geocode.tsx
import React, { useRef, useEffect, useState } from 'react';
import mapboxgl from 'mapbox-gl';

import { Geocoder } from '../src/components/Geocoder';
import { MapboxGeocoder } from '@mapbox/search-js-web';
import { useGeocodingCore } from '../src';

const ACCESS_TOKEN = MAPBOX_ACCESS_TOKEN;

const coordinatesGeocoder = async (query) => {
  // Simulate a delay to mimic an async API call
  await new Promise((resolve) => setTimeout(resolve, 250));

  // Match anything which looks like
  // decimal degrees coordinate pair.
  const matches = query.match(
    /^[ ]*(?:Lat: )?(-?\d+\.?\d*)[, ]+(?:Lng: )?(-?\d+\.?\d*)[ ]*$/i
  );
  if (!matches) {
    return null;
  }

  function coordinateFeature(lng, lat) {
    return {
      type: 'Feature',
      geometry: {
        type: 'Point',
        coordinates: [lng, lat]
      },
      properties: {
        name: 'Lng: ' + lng + ' Lat: ' + lat,
        place_formatted: 'Coordinate location',
        full_address: 'Lng: ' + lng + ' Lat: ' + lat
      }
    };
  }

  const coord1 = Number(matches[1]);
  const coord2 = Number(matches[2]);
  const geocodes = [];

  if (coord1 < -90 || coord1 > 90) {
    // must be lng, lat
    geocodes.push(coordinateFeature(coord1, coord2));
  }

  if (coord2 < -90 || coord2 > 90) {
    // must be lat, lng
    geocodes.push(coordinateFeature(coord2, coord1));
  }

  if (geocodes.length === 0) {
    // else could be either lng, lat or lat, lng
    geocodes.push(coordinateFeature(coord1, coord2));
    geocodes.push(coordinateFeature(coord2, coord1));
  }

  return geocodes;
};

export function Geocode(): React.ReactElement {
  const [map, setMap] = useState(null);
  const [value, setValue] = useState('');

  const geocodeRef = useRef(null);

  const mapContainerRef = useRef(null);

  const geocodingCore = useGeocodingCore({ accessToken: ACCESS_TOKEN });

  const runGeocodingSuggest = React.useCallback(async () => {
    const response = await geocodingCore.suggest('1600 pennsylvania ave nw');
    console.log(response);
  }, [geocodingCore]);

  useEffect(() => {
    mapboxgl.accessToken = ACCESS_TOKEN;

    const map = new mapboxgl.Map({
      container: mapContainerRef.current,
      style: 'mapbox://styles/mapbox/streets-v11',
      center: [-73.943, 40.7789],
      zoom: 11
    });

    setMap(map);

    map.addControl(new mapboxgl.FullscreenControl());

    const mapSearchControl = new MapboxGeocoder();
    mapSearchControl.accessToken = ACCESS_TOKEN;
    mapSearchControl.mapboxgl = mapboxgl;
    mapSearchControl.marker = { color: 'red' };
    mapSearchControl.componentOptions = {
      flipCoordinates: false
    };
    map.addControl(mapSearchControl);

    // geocodeRef.current.search('central park, new york ny');

    // Clean up on unmount
    return () => map.remove();
  }, []);

  return (
    <div
      style={{
        height: '800px',
        width: '100%',
        margin: '30px',
        display: 'flex',
        flexDirection: 'column'
      }}
    >
      <div style={{ margin: '24px 12px' }}>
        <h2>Geocoder component outside of a map</h2>
        <Geocoder
          accessToken={ACCESS_TOKEN}
          value={value}
          placeholder="Search for an address"
          onChange={(v) => {
            setValue(v);
            console.log('change', v);
          }}
          onSuggest={(res) => console.log('suggest', res)}
          onSuggestError={(err) => console.log('error', err)}
          onRetrieve={(res) => console.log('retrieve', res)}
          onClear={() => console.log('clear')}
          map={map}
          marker={true}
          mapboxgl={mapboxgl}
          componentOptions={{
            flipCoordinates: true
          }}
          ref={geocodeRef}
        />
      </div>
      <br></br>
      <div style={{ height: '100%', width: '100%' }}>
        <h2>Geocoder component as control to a map</h2>
        <div
          className="map-container"
          style={{ height: '100%', width: '100%' }}
          ref={mapContainerRef}
        />
      </div>
      <br></br>
      <br></br>
      <div>
        <h2>Geocoder with customSearch option</h2>
        <Geocoder
          accessToken={ACCESS_TOKEN}
          placeholder="Search for an address"
          onRetrieve={(res) =>
            console.log('retrieve with custom features', res)
          }
          map={map}
          marker={{ color: 'orange' }}
          mapboxgl={mapboxgl}
          componentOptions={{
            flipCoordinates: true,
            customSearch: coordinatesGeocoder
          }}
          ref={geocodeRef}
        />
      </div>
      <button type="button" onClick={runGeocodingSuggest}>
        useGeocodingCore Test
      </button>
    </div>
  );
}
