// @mapbox/search-js-react/web/Search.tsx
import React, { useRef, useEffect, useState } from 'react';
import mapboxgl from 'mapbox-gl';

import { SearchBox } from '../src/components/SearchBox';
import { MapboxSearchBox } from '@mapbox/search-js-web';
import { useSearchBoxCore } from '../src';

const ACCESS_TOKEN = MAPBOX_ACCESS_TOKEN;

const coordinatesGeocoder = async (query) => {
  // Match anything which looks like
  // decimal degrees coordinate pair.
  const matches = query.match(
    /^[ ]*(?:Lat: )?(-?\d+\.?\d*)[, ]+(?:Lng: )?(-?\d+\.?\d*)[ ]*$/i
  );
  if (!matches) {
    return null;
  }

  function coordinateFeature(lng, lat) {
    return {
      name: 'Lng: ' + lng + ' Lat: ' + lat,
      place_formatted: 'Coordinate location',
      full_address: 'Lng: ' + lng + ' Lat: ' + lat,
      _geometry: {
        type: 'Point',
        coordinates: [lng, lat]
      }
    };
  }

  const coord1 = Number(matches[1]);
  const coord2 = Number(matches[2]);
  const geocodes = [];

  if (coord1 < -90 || coord1 > 90) {
    // must be lng, lat
    geocodes.push(coordinateFeature(coord1, coord2));
  }

  if (coord2 < -90 || coord2 > 90) {
    // must be lat, lng
    geocodes.push(coordinateFeature(coord2, coord1));
  }

  if (geocodes.length === 0) {
    // else could be either lng, lat or lat, lng
    geocodes.push(coordinateFeature(coord1, coord2));
    geocodes.push(coordinateFeature(coord2, coord1));
  }

  return geocodes;
};

export function Search(): React.ReactElement {
  const [map, setMap] = useState(null);
  const [value, setValue] = useState('');

  const sbRef = useRef(null);

  const mapContainerRef = useRef(null);

  const searchBoxCore = useSearchBoxCore({
    accessToken: ACCESS_TOKEN
  });

  const runSearchBoxSuggest = React.useCallback(async () => {
    const response = await searchBoxCore.suggest('1600 pennsylvania ave nw', {
      sessionToken: 'test-123'
    });
    console.log(response);
  }, [searchBoxCore]);

  useEffect(() => {
    mapboxgl.accessToken = ACCESS_TOKEN;

    const map = new mapboxgl.Map({
      container: mapContainerRef.current,
      style: 'mapbox://styles/mapbox/streets-v11',
      center: [-73.943, 40.7789],
      zoom: 11
    });

    setMap(map);

    map.addControl(new mapboxgl.FullscreenControl());

    const mapSearchControl = new MapboxSearchBox();
    mapSearchControl.accessToken = ACCESS_TOKEN;
    mapSearchControl.componentOptions = {
      allowReverse: true,
      flipCoordinates: false
    };
    mapSearchControl.mapboxgl = mapboxgl;
    mapSearchControl.marker = { color: 'red' };
    map.addControl(mapSearchControl);

    // sbRef.current.search('central park, new york ny');

    // Clean up on unmount
    return () => map.remove();
  }, []);

  return (
    <div
      style={{
        height: '800px',
        width: '100%',
        margin: '30px',
        display: 'flex',
        flexDirection: 'column'
      }}
    >
      <div style={{ margin: '24px 12px' }}>
        <h2>SearchBox component outside of a map</h2>
        <SearchBox
          accessToken={ACCESS_TOKEN}
          value={value}
          placeholder="Search for an address"
          onChange={(v) => {
            setValue(v);
            console.log('change', v);
          }}
          onSuggest={(res) => console.log('suggest', res)}
          onSuggestError={(err) => console.log('error', err)}
          onRetrieve={(res) => console.log('retrieve', res)}
          onClear={() => console.log('clear')}
          map={map}
          marker={true}
          mapboxgl={mapboxgl}
          ref={sbRef}
          options={{ language: 'es', types: 'country,poi' }}
          componentOptions={{
            allowReverse: true,
            flipCoordinates: true
          }}
        />
      </div>
      <br></br>
      <div style={{ height: '100%', width: '100%' }}>
        <h2>SearchBox component as control to a map</h2>
        <div
          className="map-container"
          style={{ height: '100%', width: '100%' }}
          ref={mapContainerRef}
        />
      </div>
      <br></br>
      <br></br>
      <div style={{ margin: '24px 12px' }}>
        <h2>SearchBox with customSearch option</h2>
        <SearchBox
          accessToken={ACCESS_TOKEN}
          placeholder="Search for an address"
          onRetrieve={(res) =>
            console.log('retrieve with custom features', res)
          }
          map={map}
          marker={{ color: 'orange' }}
          mapboxgl={mapboxgl}
          ref={sbRef}
          options={{ language: 'es', types: 'country,poi' }}
          componentOptions={{
            allowReverse: true,
            flipCoordinates: true,
            customSearch: coordinatesGeocoder
          }}
        />
      </div>
      <button type="button" onClick={runSearchBoxSuggest}>
        useSearchBoxCore Test
      </button>
    </div>
  );
}
